sdk/core/allocator/alloc.h

```diff
@@ -130,53 +127,328 @@ static inline size_t leftshifted_val_msb(size_t val)
+	 */
+	SmallSize shead;
+
+	bool isPrevInUse : 1;
+	bool isCurrInUse : 1;
+
+	/* There are 30 bits free in the header here for future use */
+
+	public:
+	__always_inline auto cell_prev()
+	{
+		return displacement_proxy::
+		  Proxy<MChunkHeader, SmallSize, false, head2size, size2head>(this,
+		                                                              sprev);
+	}
+
+	__always_inline auto cell_next()
+	{
+		return displacement_proxy::
+		  Proxy<MChunkHeader, SmallSize, true, head2size, size2head>(this,
+		                                                             shead);
+	}
+
+	/**
+	 * Erase the header fields
+	 */
+	__always_inline void clear()
+	{
+		/*
+		 * This is spelled slightly oddly as using memset results in a call
+		 * rather than a single store instruction.
+		 */
+		static_assert(sizeof(*this) == sizeof(uintptr_t));
+		*reinterpret_cast<uintptr_t *>(this) = 0;
+	}
+
+	bool is_in_use()
+	{
+		return isCurrInUse;
+	}
+
+	bool is_prev_in_use()
+	{
+		return isPrevInUse;
+	}
+
+	// size of the previous chunk
+	size_t prevsize_get()
+	{
+		return head2size(sprev);
+	}
+
+	// size of this chunk
+	size_t size_get()
+	{
+		return head2size(shead);
+	}
+
+	void mark_in_use()
+	{
+		isCurrInUse              = true;
+		cell_next()->isPrevInUse = true;
+	}
+
+	void mark_free()
+	{
+		isCurrInUse              = false;
+		cell_next()->isPrevInUse = false;
+	}
+
+	/**
+	 * Land a new header somewhere within an existing chunk.
+	 *
+	 * The resulting header inherits its in-use status from this one, which
+	 * leaves consistent the "previous in use" bits of both this new header and
+	 * the successor header, but this can violate other invariants of the
+	 * system.  In particular, if this is a free chunk, then the result will be
+	 * two free chunks in a row; the caller is expected to fix this by
+	 * marking at least one of the two split chunks as in-use.
+	 */
+	MChunkHeader *split(size_t offset)
+	{
+		auto newnext = ds::pointer::offset<MChunkHeader>(this, offset);
+
+		newnext->clear();
+
+		ds::linked_list::emplace_after(this, newnext);
+		newnext->isCurrInUse = newnext->isPrevInUse = isCurrInUse;
+
+		return newnext;
+	}
+
+	/**
+	 * Build an initial pair of chunk headers.  The returned header is at the
+	 * indicated base and is suitable for installation into a free list (or
+	 * tree).  The other header is a sentinel of sorts, at the end of the
+	 * indicated span of memory.  To prevent the system from attempting to walk
+	 * beyond these bounds, ficitious "in use" bits are set true: the first
+	 * chunk claims that its predecessor is in use and the second claims that
+	 * it itself is.
+	 *
+	 * This function is marked always-inlined as it is called exactly once.
+	 */
+	__always_inline static MChunkHeader *make(void *base, size_t size)
+	{
+		size -= sizeof(MChunkHeader);
+
+		auto first = static_cast<MChunkHeader *>(base);
+		first->clear();
+		first->shead       = size2head(size);
```

Comment:
Does this overflow into the header word right after `base + size`, instead of placing the fake footer header at the end of `base + size`?

Reply:
You are right. Somehow I didn't see the subtraction.